sonic_installer/bootloader/bootloader.py

```diff
@@ -0,0 +1,70 @@
+
+import os
+import re
+import subprocess
+
+from ..common import (
+   IMAGE_DIR_PREFIX,
+   IMAGE_PREFIX,
+   default_sigpipe,
+)
+
+class Bootloader(object):
+
+    NAME = None
+    DEFAULT_IMAGE_PATH = '/tmp/sonic_image'
+
+    def get_current_image(self):
+        """returns name of the current image"""
+        with open('/proc/cmdline') as f:
+            current = re.search(r"loop=/*(\S+)/", f.read()).group(1)
+        return current.replace(IMAGE_DIR_PREFIX, IMAGE_PREFIX)
+
+    def get_next_image(self):
+        """returns name of the next image"""
+        raise NotImplementedError
+
+    def get_installed_images(self):
+        """returns list of installed images"""
+        raise NotImplementedError
+
+    def set_default_image(self, image):
+        """set default image to boot from"""
+        raise NotImplementedError
+
+    def set_next_image(self, image):
+        """set next image to boot from"""
+        raise NotImplementedError
+
+    def install_image(self, image_path):
+        """install new image"""
+        raise NotImplementedError
+
+    def remove_image(self, image):
+        """remove existing image"""
+        raise NotImplementedError
+
+    def get_binary_image_version(self, image_path):
+        """returns the version of the image"""
+        p1 = subprocess.Popen(["cat", "-v", image_path], stdout=subprocess.PIPE, preexec_fn=default_sigpipe)
+        p2 = subprocess.Popen(["grep", "-m 1", "^image_version"], stdin=p1.stdout, stdout=subprocess.PIPE, preexec_fn=default_sigpipe)
+        p3 = subprocess.Popen(["sed", "-n", r"s/^image_version=\"\(.*\)\"$/\1/p"], stdin=p2.stdout, stdout=subprocess.PIPE, preexec_fn=default_sigpipe)
+
+        stdout = p3.communicate()[0]
+        p3.wait()
+        version_num = stdout.rstrip('\n')
+
+        # If we didn't read a version number, this doesn't appear to be a valid SONiC image file
+        if not version_num:
+            return None
+
+        return IMAGE_PREFIX + version_num
+
+    def verify_binary_image(self, image_path):
+        """verify that the image is supported by the bootloader"""
+        return os.path.isfile(image_path)
+
+    @classmethod
+    def detect(cls):
+        return False
```

Comment:
How about creating a factory method in base class and create subclass after detecting. It is weird to let every subclasses detect one after another.

Reply:
Do you want something like the following ?

```
class Bootloader(object):
     @classmethod
     def get_bootloader(cls):
           return <detectedBootloader>
```

With this approch Bootloader can never have access to BOOTLOADER variable since it would be a cyclic dependency.

The only option would be to add a `register_bootloader` and have every python file call it or use a decorator on the class.

```
_bootloaders = []

def register_bootloader(cls):
    _bootloader.append(cls)

class Bootloader(object):
     @classmethod
     def get_bootloader(cls)
          return [b for b in _bootloaders if b.detect()[0]
```

But overall it's just adding complexity for something that will never scale that much.

Also there is no other way around letting each bootloader detect if they are being in used.
Otherwise you would need to move bootloader specific code outside of their own files which I believe defeats the purpose of this approach in a first place.